modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/SemanticHeaderPhase.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.phase;
+
+import org.elasticsearch.painless.lookup.PainlessLookup;
+import org.elasticsearch.painless.node.SClass;
+import org.elasticsearch.painless.node.SFunction;
+import org.elasticsearch.painless.symbol.FunctionTable;
+import org.elasticsearch.painless.symbol.ScriptScope;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SemanticHeaderPhase extends UserTreeBaseVisitor<ScriptScope, Void> {
+
+    @Override
+    public Void visitClass(SClass userClassNode, ScriptScope scriptScope) {
+        for (SFunction userFunctionNode : userClassNode.getFunctionNodes()) {
+            visitFunction(userFunctionNode, scriptScope);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Void visitFunction(SFunction userFunctionNode, ScriptScope scriptScope) {
+        String functionName = userFunctionNode.getFunctionName();
+        List<String> canonicalTypeNameParameters = userFunctionNode.getCanonicalTypeNameParameters();
+        List<String> parameterNames = userFunctionNode.getParameterNames();
+        int parameterCount = canonicalTypeNameParameters.size();
+
+        if (parameterCount != parameterNames.size()) {
```

Comment:
Check `canonicalTypeNameParameters.size()` != `parameterNames.size()`?

Reply:
Good catch. Guessing this was left over from prior to the refactor where this was actually re-used more than once. Fixed.